pkg/daemon/certs/certs.go

```diff
@@ -0,0 +1,198 @@
+	scanner := bufio.NewScanner(f)
+	scanner.Split(splitAtPEMCert())
+	var certs []*x509.Certificate
+	for scanner.Scan() {
+		// Read next token
+		pemData := scanner.Text()
+		block, _ := pem.Decode([]byte(pemData))
+		if block == nil {
+			return nil, fmt.Errorf("No PEM data read from token")
+		}
+		// Decode and parse the certificate
+		cert, err := x509.ParseCertificate(block.Bytes)
+		if err != nil {
+			return nil, fmt.Errorf("Failed to parse certificate: %v\n", err)
+		}
+		certs = append(certs, cert)
+	}
+	return certs, nil
+}
+
+// splitAtPEMCert is a custom closure function to split data into complete PEM-encoded certificates
+func splitAtPEMCert() func(data []byte, atEOF bool) (advance int, token []byte, err error) {
+	const endCertTag = "-----END CERTIFICATE-----"
+	searchBytes := []byte(endCertTag)
+	searchLen := len(searchBytes)
+	return func(data []byte, atEOF bool) (advance int, token []byte, err error) {
+		dataLen := len(data)
+		if atEOF && dataLen == 0 {
+			// Empty file, do nothing
+			return 0, nil, nil
+		}
+
+		// Find next separator and return the token
+		if i := bytes.Index(data, searchBytes); i >= 0 {
+			return i + searchLen, data[0 : i+searchLen], nil
+		}
+
+		if atEOF {
+			return dataLen, data, fmt.Errorf("Hit end of file without finding terminating separator %s", endCertTag)
+		}
+		// Otherwise, continue reading file data
+		return 0, nil, nil
+	}
+}
+
+// ensureCertsAreImported imports the expected certificates into the instance's root system trust store, if not already
+// present. Returns a boolean if any certificates were imported
+func ensureCertsAreImported(expectedCerts []*x509.Certificate) (bool, error) {
+	// Open the root certificate store
+	systemStore, err := windows.CertOpenStore(windows.CERT_STORE_PROV_SYSTEM, 0, 0,
+		windows.CERT_SYSTEM_STORE_LOCAL_MACHINE, uintptr(unsafe.Pointer(syscall.StringToUTF16Ptr("ROOT"))))
+	if err != nil {
+		return false, fmt.Errorf("Failed to open root certificate store: %v", err)
+	}
+	defer func() {
+		if err := windows.CertCloseStore(systemStore, 0); err != nil {
+			klog.Error("Failed to close root system certificate store")
+		}
+	}()
+
+	// Get all existing from system store
+	existingCerts, err := getAllCerts(systemStore)
+	if err != nil {
+		return false, fmt.Errorf("Failed to read certificate file: %v", err)
+	}
+
+	certChange := false
+	for _, cert := range expectedCerts {
+		if containsCert(existingCerts, cert) {
+			// Cert already exists as expected, do nothing
+			continue
+		}
+
+		// Add the certificate to the instance's system-wide trust store
+		if err := addCertToStore(systemStore, cert); err != nil {
+			return false, err
+		}
+		certChange = true
+	}
+	return certChange, nil
+}
+
+// getAllCerts reads all the certs from the given certificate store
+func getAllCerts(store windows.Handle) ([]*x509.Certificate, error) {
+	var existingSystemCerts []*x509.Certificate
+	var certContext *windows.CertContext
+	var err error
```

Comment:
Consider declaring the error and the list as return parameters in the function's signature.